### PR TITLE
Reduce MoE peak memory in unfused EP combine path

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Reduces transient memory usage in MoE expert output assembly and EP token combine by avoiding full intermediate list/cat buffers and chunking the return all-to-all/unpermute path.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [X] I have added relevant unit tests
- [X] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

This PR targets peak memory in the unfused MoE all-to-all dispatcher path:

- Preallocates `SequentialMLP` local expert output and copies each expert result into the final buffer instead of storing per-expert outputs and `torch.cat`-ing them.
- Replaces the unfused `sort_chunks_by_idxs` split/cat fallback with a custom autograd reorder that copies into a preallocated output and scatters gradients back without retaining chunked source views.
- Adds a chunked EP combine path that performs return `all_to_all_single` and final unpermute by hidden-dimension chunks, reducing the full-size temporary allocation during token combine.
- Keeps the chunked path gated to the supported case: `ep_size > 1`, no `drop_and_pad`, no `moe_permute_fusion`, and no shared experts. Existing paths are preserved otherwise.

Large MoE runs can hit OOM from transient buffers in the combine side of expert parallelism, especially when the full all-to-all output and unpermuted output are both live. This change lowers that peak by streaming the combine/unpermute work through bounded chunks.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 1]
```diff
@@ -845,6 +982,26 @@ def token_combine(
...
 
+    def _can_chunk_ep_combine_unpermute(self) -> bool:
+        return (
+            self.ep_size > 1
+            and self.input_splits is not None
+            and self.output_splits is not None
+            and not self.drop_and_pad
+            and not self.config.moe_permute_fusion
+            and self.shared_experts is None
+        )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply megatron/core/transformer/moe/token_dispatcher.py:1054]: I changed the chunked combine path to explicit opt-in via --moe-chunked-ep-combine, which defaults to False. Existing runs therefore keep the original single-all-to-all path and see no additional collective or launch overhead.

I also replaced the hardcoded 8 MiB value with --moe-chunked-ep-combine-max-chunk-bytes, including positive-value validation. I chose explicit opt-in rather than a heuristic threshold because the memory/throughput tradeoff is workload-dependent.

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 2]
```diff
@@ -50,6 +50,139 @@
...
+            send = hidden_states.narrow(1, start, width).contiguous()
+            recv = hidden_states.new_empty((output_rows, width))
+            torch.distributed.all_to_all_single(
+                recv,
+                send,
+                output_split_sizes=output_split_sizes,
+                input_split_sizes=input_split_sizes,
+                group=group,
+            )
+            output_tokens.narrow(1, start, width).index_add_(0, sorted_indices, recv)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply megatron/core/transformer/moe/token_dispatcher.py:186]: Yes, although index_add_ itself is intentional, the complete chunked operation—including its Python chunk loop—should not be treated as graph-capturable.

The path is now disabled whenever CUDA graphs are configured and also checks not is_graph_capturing() at runtime. In those cases it falls back to the existing all-to-all plus unpermute path. The test now verifies that active graph capture makes the chunked path ineligible.

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 3]
```diff
@@ -50,6 +50,139 @@
...
+            return tuple(int(x) for x in values.tolist())
+        return tuple(int(x) for x in values)
+
+    @staticmethod
+    def _max_rows(group, local_rows: int, device: torch.device) -> int:
+        if group.size() == 1:
+            return int(local_rows)
+        rows = torch.tensor([int(local_rows)], device=device, dtype=torch.long)
+        torch.distributed.all_reduce(rows, op=torch.distributed.ReduceOp.MAX, group=group)
+        return int(rows.item())
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply megatron/core/transformer/moe/token_dispatcher.py]: Removed the _max_rows all_reduce and GPU .item() entirely.

During preprocessing, every rank already has the gathered routing counts with shape [tp, source_ep, num_experts]. For the current TP rank, I collapse this into the complete [source_ep, destination_ep] split matrix. The maximum source-row and destination-column totals provide an exact, rank-consistent bound for both combine input and output rows.

The resulting scalar is copied to CPU alongside the existing split metadata using the existing D2H stream and synchronization event, so this adds neither a collective nor a new synchronization point.

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 4]
```diff
@@ -55,6 +56,173 @@
...
+            f"{tuple(tokens_per_expert.shape)} for ep_size={ep_size}"
+        )
+    if tokens_per_expert.size(1) % ep_size != 0:
+        raise RuntimeError(
+            f"num_experts={tokens_per_expert.size(1)} must be divisible by ep_size={ep_size}"
+        )
+
+    split_matrix = tokens_per_expert.reshape(ep_size, ep_size, -1).sum(dim=-1)
+    max_input_rows = split_matrix.sum(dim=0).max()
+    max_output_rows = split_matrix.sum(dim=1).max()
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply megatron/core/transformer/moe/token_dispatcher.py]: addressed

**Review by yaox12 — APPROVED** [review]

**Comment by kvareddy** [comment]
Why do we need these changes? Can you provide supporting data and perf implications?

**Comment by HollowMan6** [comment]
Thank you for your review, this was a part of the changes when I tried to tackle moe memory issue, now I got some more experiment data and it looks like it's better to just enable `moe_permute_fusion` instead, as in any case perf will not be better than with `moe_permute_fusion` enabled, so closing this.
